deadline_/social/models.py

```diff
@@ -0,0 +1,384 @@
+    newsfeed_item = models.ForeignKey(NewsfeedItem, related_name='likes')
+    author = models.ForeignKey(User)
+    unique_together = ('newsfeed_item', 'author')
+
+
+class NotificationManager(models.Manager):
+    """
+    Use a custom Notification manager for specific type of Notification creation
+    """
+    def create(self, *args, **kwargs):
+        raise ForbiddenMethodError('This method is not intended to be used. '
+                                   'Please use any of the create_xxx methods to create a specific type of notification')
+
+    def _create(self, *args, **kwargs):
+        """
+        The create() method, intentionally marked as private as it is not intended for usage at all,
+            except for testing purposes
+        """
+        return super().create(*args, **kwargs)
+
+    def create_receive_follow_notification(self, recipient: User, follower: User):
+        """
+        Creates a Notification that a User has been followed
+        ex: Stanislav has followed you!
+        """
+        if follower == recipient:
+            raise InvalidFollowError('You cannot follow yourself!')
+        return self._create(recipient=recipient, type=RECEIVE_FOLLOW_NOTIFICATION,
+                            content={
+                                'follower_id': follower.id,
+                                'follower_name': follower.username
+                            })
+
+    def create_receive_submission_upvote_notification(self, submission: 'Submission', liker: User):
+        """
+        Creates a Notification that a User has liked your submission
+        ex: Stanislav has liked your Submission for Challenge Robert's Pass
+        """
+        if liker == submission.author:
+            return
+
+        return self._create(recipient=submission.author, type=RECEIVE_SUBMISSION_UPVOTE_NOTIFICATION,
+                            content={
+                                'submission_id': submission.id,
+                                'challenge_id': submission.challenge.id,
+                                'challenge_name': submission.challenge.name,
+                                'liker_id': liker.id,
+                                'liker_name': liker.username
+                            })
+
+    def create_receive_nw_item_like_notification(self, nw_item: NewsfeedItem, liker: User):
+        if liker == nw_item.author:
+            return
+
+        return self._create(recipient=nw_item.author, type=RECEIVE_NW_ITEM_LIKE_NOTIFICATION,
+                            content={'nw_content': nw_item.content,
+                                     'liker_id': liker.id, 'liker_name': liker.username, 'nw_type': nw_item.type})
+
+    def create_new_challenge_notification(self, recipient: User, challenge: 'Challenge'):
+        """ A notification which notifies the user that a new challenge has appeared on the site"""
+        # TODO: Figure out a way to add new challenges and create notifications. Maybe add a created_on
```

Comment:
Do we need this at this point or we can leave it as is for now?

Reply:
I don't know if we *need* this, but it'd be nice to have.
I have come up with a solution: it's to add a `created_on`  field to every challenge fixture and when we create the fixtures, `Challenge.objects.create` gets called, so we can attach a hook on creation to check if the `Challenge` is new enough to notify people. Then, we'll also need a way to mark this challenge as having sent its notifications (maybe another table as it makes no sense to put in the main one, since it'll get queried very rarely), so we don't happen to send them twice

Reply:
@nemalp Thinking about my comment above, I think it doesn't make sense to check the notifications on challenge creation and this bears the question - why do we need fixtures at all?
We could just have a DB backup which stores the challenges we want.
Regardless:

The main question is: __How will we create new challenges?__
We still don't support creating new challenges apart from writing a new fixture and loading them all, which obviously shouldn't be the way. Maybe we should actually create some Challenge creation page for admins only?

Going back to the challenge notifications, I figured that the logic for creating a notification (and sending it) for __every__ user should be in the `notifications` app and that message (the message telling it to create a notif for this new challenge) should come through RabbitMQ.
I'm thinking of implementing that in `notifications` now and leaving out the logic that actually _sends the message_ for later, when we figure out what to do.

@nemalp Any ideas/what do you think?

Reply:
@Enether 

Implementing a challenge creation page seems the right way to go and we should do it at some point. Perhaps, we can do that along with the education system. Obviously we're going to need something similar there. What do you think?

Reply:
@nemalp Could be, it just means we'd need . to stay without "New Challenge Notifications" until then

Reply:
I am fine with that. Let's skip it for now if you are fine too.